src/stats/stats.jl

```diff
@@ -0,0 +1,144 @@
+export ExecutionStats, statshead, statsline
+
+const STATUS = Dict(:unknown => "unknown",
+                    :first_order => "first-order stationary",
+                    :max_eval => "maximum number of function evaluations",
+                    :max_time => "maximum elapsed time",
+                    :max_iter => "maximum iteration",
+                    :neg_pred => "negative predicted reduction",
+                    :unbounded => "objective function may be unbounded from below",
+                    :exception => "unhandled exception",
+                    :stalled => "stalled"
+                   )
+
+type ExecutionStats
+  status :: Symbol
+  solution :: Vector # x
+  obj :: Float64 # f(x)
+  dual_feas :: Float64 # ‖∇f(x)‖₂ for unc, ‖P[x - ∇f(x)] - x‖₂ for bnd, etc.
+  iter :: Int
+  eval :: NLPModels.Counters
+  elapsed_time :: Float64
+  solver_specific :: Dict{Symbol,Any}
+end
+
+function ExecutionStats{T}(status :: Symbol;
+                           x :: Vector=Float64[], f :: Float64=Inf,
+                           normg :: Float64=Inf, iter :: Int=-1,
+                           t :: Float64=Inf, eval :: NLPModels.Counters=Counters(),
+                           solver_specific :: Dict{Symbol,T}=Dict{Symbol,Any}(),
+                           kwargs...)
+  if !(status in keys(STATUS))
+    s = join(keys(STATUS, ", "))
+    error("$status is not a valid status. Use one of the following: $s")
+  end
+  for (k,v) in kwargs
+    if k == :solution || k == :sol
+      x = v
+    elseif k == :objective || k == :obj
+      f = v
+    elseif k == :dual_feas || k == :opt
+      normg = v
+    elseif k == :elapsed_time || k == :time
+      t = v
+    else
+      throw(UndefVarError(k))
+    end
+  end
+  return ExecutionStats(status, x, f, normg, iter, deepcopy(eval), t, solver_specific)
+end
+
+import Base.show, Base.print, Base.println
+
+function show(io :: IO, stats :: ExecutionStats)
+  show(io, "Execution stats: $(getStatus(stats))")
+end
+
+# TODO: Expose NLPModels dsp in nlp_types.jl function print
+function disp_vector(io :: IO, x :: Vector)
+  if length(x) == 0
+    @printf(io, "∅")
+  elseif length(x) <= 5
+    Base.show_delim_array(io, x, "", " ", "", false)
+  else
+    Base.show_delim_array(io, x[1:4], "", " ", "", false)
+    @printf(io, " ⋯ %s", x[end])
+  end
+end
+
+function print(io :: IO, stats :: ExecutionStats; showvec :: Function=disp_vector)
+  # TODO: Show evaluations
+  @printf(io, "Execution stats\n")
+  @printf(io, "  status: "); show(io, getStatus(stats)); @printf(io, "\n")
+  @printf(io, "  objective value: "); show(io, stats.obj); @printf(io, "\n")
+  @printf(io, "  dual feasibility: "); show(io, stats.dual_feas); @printf(io, "\n")
+  @printf(io, "  solution: "); showvec(io, stats.solution); @printf(io, "\n")
+  @printf(io, "  iterations: "); show(io, stats.iter); @printf(io, "\n")
+  @printf(io, "  elapsed time: "); show(io, stats.elapsed_time); @printf(io, "\n")
+  if length(stats.solver_specific) > 0
```

Comment:
Maybe this should be delegated to the solver-specific stats type?

Reply:
Maybe `ExecutionStats` should be `GenericExecutionStats <: AbstractExecutionStats`?